server/src/push/emitters.js

```diff
@@ -0,0 +1,63 @@
+import sleep from '../utils';
+import { sendPush } from './index';
+
+const getPushEmitters = ({ models }) => {
+  const {
+    Group,
+  } = models;
+
+  const emitters = {
+    async sendTestPush({ devices, message, delay }) {
+      if (delay) {
+        await sleep(5000);
+      }
+      return sendPush({ devices, message });
+    },
+
+    pushScheduleToGroup(schedule) {
+      Group.findById(schedule.groupId).then((group) => {
+        group.getUsers().then((users) => {
+          const results = [];
+          return Promise.all(
+            users.map(usr =>
+              usr.getDevices().then((dvc) => {
+                // unpack the array into the new array
+                results.push(...dvc);
+              }),
+            ),
+          ).then(() => {
+            sendPush({
+              devices: results,
+              message: `Schedule #${schedule.id} Created\n${schedule.name}`,
+            });
+          });
+        });
+      });
+    },
+    pushEventToGroup(event) {
+      Group.findById(event.groupId).then((group) => {
```

Comment:
i'm fairly certain you can traverse these relationships in one query using sequelize. Look up how to a join... You should be able to get a complete list of devices given an event's groupId with only one roundtrip. If it's too much effort don't worry - premature optimisation is a waste of time. Stating intentions with a comment or TODO can help :)

Reply:
Very possible im just worried ill spend 2 days fighting with sequalizer in the process